source/common/network/connection_impl.cc

```diff
@@ -540,9 +478,48 @@ ClientConnectionImpl::ClientConnectionImpl(
     Event::Dispatcher& dispatcher, const Address::InstanceConstSharedPtr& remote_address,
     const Network::Address::InstanceConstSharedPtr& source_address,
     Network::TransportSocketPtr&& transport_socket)
-    : ConnectionImpl(dispatcher, remote_address->socket(Address::SocketType::Stream),
-                     remote_address, nullptr, source_address, std::move(transport_socket), false,
-                     false) {}
+    : ConnectionImpl(dispatcher, ConnectionSocketPtr{new ClientSocketImpl(remote_address)},
+                     std::move(transport_socket), false) {
+  if (source_address != nullptr) {
+    int rc = source_address->bind(fd());
+    if (rc < 0) {
+      ENVOY_LOG_MISC(debug, "Bind failure. Failed to bind to {}: {}", source_address->asString(),
+                     strerror(errno));
+      // Set a special error state to ensure asynchronous close to give the owner of the
+      // ConnectionImpl a chance to add callbacks and detect the "disconnect"
+      bind_error_ = true;
+
+      // Trigger a write event to close this connection out-of-band.
+      file_event_->activate(Event::FileReadyType::Write);
+    }
+  }
+}
+
+void ClientConnectionImpl::connect() {
+  ENVOY_CONN_LOG(debug, "connecting to {}", *this, socket_->remoteAddress()->asString());
+  int rc = socket_->remoteAddress()->connect(fd());
+  if (rc == 0) {
+    // write will become ready.
+    ASSERT(connecting_);
+  } else {
+    ASSERT(rc == -1);
+    if (errno == EINPROGRESS) {
+      ASSERT(connecting_);
+      ENVOY_CONN_LOG(debug, "connection in progress", *this);
+    } else {
+      // read/write will become ready.
+      immediate_connection_error_ = true;
+      connecting_ = false;
+      ENVOY_CONN_LOG(debug, "immediate connection error: {}", *this, errno);
+    }
+  }
+
+  // The local address can only be retrieved for IP connections. Other
+  // types, such as UDS, don't have a notion of a local address.
+  if (socket_->remoteAddress()->type() == Address::Type::Ip) {
+    dynamic_cast<ClientSocket*>(socket_.get())->setLocalAddress(Address::addressFromFd(fd()));
```

Comment:
not super crazy about this dynamic cast. I honestly don't recall at this point why I wrote this code the way I did (with connect/bind in the base class vs. client class). Probably to deal with some of the issues that you are dealing with here. What was the rationale for splitting it out? I agree it's better to have this be part of the client class, but then we face some of these cast issues. Just curious on your thought process here (I'm not sure if right solution). Instead of dynamic cast should we just allow changing local address on connection socket?

Reply:
Splitting the functionality was not the reason for the dynamic cast, but the hierarchy of the new Socket classes was. I refactored the abstract Socket classes back to one (ConnectionSocket) removing the need for the dynamic cast. There was no need for the hierarchy now that setLocalAddress and resetLocalAddress are combined to the same method.